From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <nonplay@bxteam.org>
Date: Sun, 6 Jul 2025 02:23:03 +0300
Subject: [PATCH] lithium: faster hash palette

This patch is based on the following mixins and classes:
* "net/caffeinemc/mods/lithium/mixin/chunk/palette/StrategyMixin.java"
By: 2No2Name <2No2Name@web.de>
As part of: Lithium (https://github.com/CaffeineMC/lithium)
Licensed under: LGPL-3.0 (https://www.gnu.org/licenses/lgpl-3.0.html)

Replaces the vanilla hash palette with an optimized variant

diff --git a/net/minecraft/world/level/chunk/HashMapPalette.java b/net/minecraft/world/level/chunk/HashMapPalette.java
index fde0a0e4b2a3b2bd4f2ac893703c70dd55b1ddc2..fc53bb226c0e1b6114d26ffcbd9a79dcd658eba7 100644
--- a/net/minecraft/world/level/chunk/HashMapPalette.java
+++ b/net/minecraft/world/level/chunk/HashMapPalette.java
@@ -26,6 +26,14 @@ public class HashMapPalette<T> implements Palette<T> {
         this.values = values;
     }
 
+    // DivineMC start - lithium: faster hash palette
+    @SuppressWarnings("DataFlowIssue")
+    protected HashMapPalette(int bits, boolean dummy) {
+        this.bits = bits;
+        this.values = dummy ? null : CrudeIncrementalIntIdentityHashBiMap.create((1 << bits) + 1);
+    }
+    // DivineMC end - lithium: faster hash palette
+
     public static <A> Palette<A> create(final int bits, final List<A> paletteEntries) {
         return new HashMapPalette<>(bits, paletteEntries);
     }
diff --git a/net/minecraft/world/level/chunk/Strategy.java b/net/minecraft/world/level/chunk/Strategy.java
index 38e4abe3307c020ec7e2385a559ff4708b09c899..b454683fc3d1b26bceac990ce3046f2c0c94d62a 100644
--- a/net/minecraft/world/level/chunk/Strategy.java
+++ b/net/minecraft/world/level/chunk/Strategy.java
@@ -4,6 +4,14 @@ import net.minecraft.core.IdMap;
 import net.minecraft.util.Mth;
 
 public abstract class Strategy<T> {
+    // DivineMC start - lithium: faster hash palette
+    private static final Palette.Factory HASH = net.caffeinemc.mods.lithium.common.world.chunk.LithiumHashPalette::create;
+    private static final Configuration HASH_4BIT = new Configuration.Simple(HASH, 4);
+    private static final Configuration HASH_5BIT = new Configuration.Simple(HASH, 5);
+    private static final Configuration HASH_6BIT = new Configuration.Simple(HASH, 6);
+    private static final Configuration HASH_7BIT = new Configuration.Simple(HASH, 7);
+    private static final Configuration HASH_8BIT = new Configuration.Simple(HASH, 8);
+    // DivineMC end - lithium: faster hash palette
     public static final Palette.Factory SINGLE_VALUE_PALETTE_FACTORY = SingleValuePalette::create; // Paper - Anti-Xray
     private static final Palette.Factory LINEAR_PALETTE_FACTORY = LinearPalette::create;
     private static final Palette.Factory HASHMAP_PALETTE_FACTORY = ca.spottedleaf.moonrise.patches.fast_palette.FastHashMapPalette::create; // Paper - optimise palette reads
@@ -36,11 +44,14 @@ public abstract class Strategy<T> {
             public Configuration getConfigurationForBitCount(final int entryBits) {
                 return switch (entryBits) {
                     case 0 -> Strategy.ZERO_BITS;
-                    case 1, 2, 3, 4 -> Strategy.FOUR_BITS_LINEAR;
-                    case 5 -> Strategy.FIVE_BITS_HASHMAP;
-                    case 6 -> Strategy.SIX_BITS_HASHMAP;
-                    case 7 -> Strategy.SEVEN_BITS_HASHMAP;
-                    case 8 -> Strategy.EIGHT_BITS_HASHMAP;
+                    // DivineMC start - lithium: faster hash palette
+                    case 1, 2 -> Strategy.FOUR_BITS_LINEAR;
+                    case 3, 4 -> Strategy.HASH_4BIT;
+                    case 5 -> Strategy.HASH_5BIT;
+                    case 6 -> Strategy.HASH_6BIT;
+                    case 7 -> Strategy.HASH_7BIT;
+                    case 8 -> Strategy.HASH_8BIT;
+                    // DivineMC end - lithium: faster hash palette
                     default -> new Configuration.Global(this.globalPaletteBitsInMemory, entryBits);
                 };
             }
